rfd/0048-war-dialler-node-tracker.md

```diff
@@ -0,0 +1,199 @@
+
+```
+	Original behaviour:
+                       ┌──────┐
+   ┌───connection─────>│proxy1│<──reverse-tunnel──┐
+   │                   └──────┘                   │
+┌──┴───┐                                        ┌─┴──┐
+│client│                                        │node│
+└──┬───┘                                        └─┬──┘
+   │                   ┌──────┐                   │
+   └───connection─────>│proxy2│<──reverse-tunnel──┘
+                       └──────┘
+	A client can reach the requested node through any proxy because the node has a reverse tunnel open to all proxies in a cluster.
+
+
+
+    New behaviour:
+                         ┌──────┐         ┌────────────┐
+   ┌─(1)─connection-1───>│proxy1├──(2)───>│node tracker│
+   │                     └──╥───┘         └────────────┘
+┌──┴───┐                    ║
+│client│               (3) tunnel
+└──┬───┘                    ║
+   │                     ┌──╨───┐                    ┌────┐
+   └────connection-2────>│proxy2│<───reverse-tunnel──┤node│
+                         └──────┘                    └────┘
+
+	Case 1 - connection 1:
+	1- A client connects to proxy-1 that can't reach the requested node.
+	2- Proxy-1 sends a lookup request to the node tracker asking which proxy is connected to the requested node. The node tracker replies with the address of proxy-2.
+	3- Proxy-1 opens a tunnel (RFD 49?) to proxy-2 and asks proxy-2 to dial the requested node.
+
+	Case 2 - connection 2:
+	nothing changes when a client connects to a proxy that has got a direct connection to the requested node.
+```
+
+### Scope
+
+The first iteration of the node tracker service will be a barebones implementation of an in memory cache represented by a simple golang map.
+
+### How it works
+
+The node tracker service runs a grpc server and exposes three endpoints to teleport proxies:
+- AddNode: which takes a `node ID`, `proxy ID` and `proxy address` as parameters among others. A Proxy will call this endpoint when it receives a heartbeat request from a node agent connected via a reverse tunnel, signaling to the node tracker that the node can be reached through this proxy, thus creating a node-to-proxy relationship.
+- RemoveNode: which takes a `node ID` as a parameter. A proxy will call this endpoint when it fails to receive heartbeat requests from a node agent after a certain threshold.
+- GetProxies: which takes `node ID` as a parameter. A proxy will call this endpoint when looking up which proxy is connected to a node it doesn't have a direct connection to. The node tracker will reply with the addresses of proxies that can reach the node.
+
+
+The service also runs a background job responsible for cleaning stale relationships that haven't been updated before a certain threshold.
+
+### Limits
+
+Currently we only support running one instance of a node tracker and because of that we're missing key features usually found in distributed systems, such as:
```

Comment:
See my comment above. Having two trackers embedded within proxies will simplify design (no need for GRPC etc) and resiliency.